### PR TITLE
[uss_qualifier] rework oir_simple: check subscription extent needs to cover OIR

### PR DESCRIPTION
**Pull request by Shastick**
Extend the OIRSimple scenario with two test cases that verify that the DSS properly validates explicit subscriptions being attached to an OIR upon creation or mutation.

(This covers situation 0 and 3 from [this DSS issue](https://github.com/interuss/dss/issues/1088))

This PR sits on top of #1009, please only review the last commit

### CODE REVIEW

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 1]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
...
+it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../requirements/astm/f3548/v21.md)**
+
 ### [Create an operational intent reference test step](./fragments/oir/crud/create_query.md)
 
-Create an operational intent reference to be used in this scenario.
+When the provided subscription covers the extent of the OIR, the OIR can be created.
+
+### [OIR is attached to expected subscription test step](./fragments/oir/oir_has_expected_subscription.md)
+
+This step verifies that the OIR is attached to the subscription provided upon creation.
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:54]: Looks like those two steps are actually a single one: include one of them as a fragment.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 2]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
...
+
+### [Create a subscription test step](./fragments/sub/crud/create_query.md)
+
+Create an additional explicit subscription to be used in this test case.
+
+### Attempt to replace OIR's existing explicit subscription with an insufficient one test step
+
+This step verifies that an OIR's existing explicit subscription cannot be replaced with an explicit subscription that does not cover the extent of the OIR.
+
+#### 🛑 Request to mutate OIR while providing an incorrect subscription fails check
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:69]: Can you be more specific than 'incorrect' in the check title?

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 3]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
...
+
+If the DSS under test allows the qualifier to replace an OIR's existing explicit subscription with an explicit subscription that does not cover the extent of the OIR,
+it is in violation of **[astm.f3548.v21.DSS0005,1](../../../../requirements/astm/f3548/v21.md)**
+
+### [OIR is attached to expected subscription test step](./fragments/oir/oir_has_expected_subscription.md)
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:74]: Sounds like this shouldn't be a step but rather included as a fragment under the above step.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 4]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
...
+
+This step verifies that the OIR is still attached to the previous, valid, subscription.
+
+### [Replace the OIR's explicit subscription test step](./fragments/oir/crud/update_query.md)
+
+This step verifies that an OIR attached to an explicit subscription can be mutated in order to be attached
+to another explicit subscription that properly covers the extent of the OIR.
+
+### [OIR is attached to expected subscription test step](./fragments/oir/oir_has_expected_subscription.md)
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:83]: Sounds like this shouldn't be a step but rather included as a fragment under the above step.

**Review by mickmis — COMMENTED** [review]
Note: I've only reviewed the .md so far.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 5]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
 
 This step ensures that no entities with the known test IDs exists in the DSS.
 
+### [Create a subscription test step](./fragments/sub/crud/create_query.md)
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:31]: The first case does not need to run that step, as it uses the subscription created in the setup case.

I tried to run the setup test case multiple times as part of some cleanup between test cases, but the logic won't allow to run it multiple times.

Possibly I could define a "reset workspace" test case instead and run it multiple times if that is allowed?

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md [hunk 6]
```diff
@@ -28,14 +28,74 @@ Verifies the behavior of a DSS for simple interactions pertaining to operational
...
+
+### [OIR is attached to expected subscription test step](./fragments/oir/oir_has_expected_subscription.md)
+
+This step verifies that the OIR is attached to the subscription provided upon mutation.
 
 ## Deletion requires correct OVN test case
 
 Ensures that a DSS will only delete OIRs when the correct OVN is presented.
 
+### [Ensure clean workspace test step](./clean_workspace.md)
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:91]: The intention is mainly to reset entities to a clearly known state. Some thoughts here:
 - I don't think it is wise to split things to separate scenarios as soon as we have "cleanup" needs between test cases: scenarios such as OIRSimple may end up doing many variants of the same operation to check various corner cases, and separating each corner case to its own scenario seems counter-productive.
 - if test cases contain steps to do some resetting/cleanups, I think it is worth having those clearly documented, and a dedicated step seems ideal for that? 

> that may not be required an just specific entities may need to be removed.

One thing that could be done to have a more specific cleanup is to just remove the known entities without doing the search. It will still require its own step though, which is why I went with the option of just re-cleaning and re-setting the workspace.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md (hunk 1 above)
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:54]: Created a new `oir_has_expected_subscription` fragment which is now part of the "Create an operational intent reference" test step

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:87]: As a sidenote: we already do such cleaning at the beginning of some test cases in the [implicit subscription handling scenario](https://github.com/interuss/monitoring/blob/5cfb1ca53ee47c9b29d2d55e397e577a3de44a90/monitoring/uss_qualifier/scenarios/astm/utm/dss/oir_implicit_sub_handling.md#ensure-clean-workspace-test-step-1), for example.

**Review by mickmis — CHANGES_REQUESTED** [review]
TBH I'm having a hard time understanding the scenario and I end up spending too much time doing that. But I need to understand the intent of the test and determine if it tests what it claims to test to properly review it. That also means it will also be difficult for users to understand the scenario when they will use it.
Notably it is difficult to understand for what purpose and why entities are created, and what's special about them. Naming them and highlighting their properties would help. It would be great if you could keep it in mind as well for future scenarios as I expect that we will have to produce many of them in the short-term. Thanks!

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:31]: The subscription is needed in _some_ steps.

From our back and forth so far and your other comments, I expect that you would prefer a solution where:

 - the setup test case is limited to cleaning up entities that might have been left over from an earlier scenario, nothing else
 - each test case precisely creates and removes the entities it needs

Can you confirm this would work for you?

[the remark this answers is not in the document]
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/op_intent_ref_simple.md:31]: > the setup test case is limited to cleaning up entities that might have been left over from an earlier scenario, nothing else

Not necessarily: ground preparation for the rest of the scenario (without actually testing stuff relevant to the scenario) is fine, made for that even. There was just:
- confusion on my end regarding to why this subscription is created here, naming it and specifying its properties help
- intent to scope creation of things close to where they are needed, but they are needed in multiple cases so that works

**Comment by Shastick** [comment]
@mickmis I've split out https://github.com/interuss/monitoring/pull/1009 so as to switch to the new convention: the plan is to create and delete OIRs (and subscriptions) within most of the test cases rather than once at the beginning and end of the scenario

**Comment by Shastick** [comment]
Closing this PR, will open a clean one for clarity
